### ISSUE TITLE
How to build this project?

### ISSUE THREAD

**Issue by reitzig** [Kotlin/kotlinx.serialization#297]
I tried opening a checkout of this repo in IDEA, but all I got from Gradle is this:

> Cause: sun.misc.Unsafe.defineClass(Ljava/lang/String;[BIILjava/lang/ClassLoader;Ljava/security/ProtectionDomain;)Ljava/lang/Class;

This is with IDEA-U 2018.3.1.

Do I have to configure something differently?

![image](https://user-images.githubusercontent.com/4246780/49609013-92531b80-f99a-11e8-8bbf-173072a88a11.png)

Changing to the built-in Java 10:

> Invalid Gradle JDK configuration found. Open Gradle Settings 
Platform SDK does not point to valid JDK (/home/raphael/Software/apps/IDEA-U/ch-0/183.4588.61)

**Comment by reitzig** [comment]
It seems that the included Gradle wrapper only runs with Java 8. With this selection for Gradle, it works:

![image](https://user-images.githubusercontent.com/4246780/49609163-04c3fb80-f99b-11e8-9de3-d946e87f5ed2.png)

**Comment by sandwwraith** [comment]
Gradle 4.7 is required to build this project, and as far as I know, it runs correctly only on Java 8 (or maybe 9)

**Comment by reitzig** [comment]
Okay, that sorted, I'm trying to run a test but get:

```
Information:Builder "Kotlin Builder" requested rebuild of module chunk "org.jetbrains.kotlinx.common.main"
Information:Builder "Kotlin Builder" requested rebuild of module chunk "org.jetbrains.kotlinx.common.test"
Information:Kotlin: kotlinc-jvm 1.3.11 (JRE 1.8.0_181-b13)
Information:javac 1.8.0_181 was used to compile java sources
Information:Modules "org.jetbrains.kotlinx.jvm.main", "org.jetbrains.kotlinx.jvm.test" were fully rebuilt due to project configuration/dependencies changes
Information:10.12.18 08:03 - Compilation completed with 15 errors and 0 warnings in 23 s 526 ms
/home/raphael/Daten/Code/kotlinx.serialization/runtime/jvm/src/test/kotlin/kotlinx/serialization/formats/RandomTests.kt
Error:(29, 38) Kotlin: Unresolved reference: proto
Error:(57, 64) Kotlin: Unresolved reference: TestInt32
Error:(66, 64) Kotlin: Unresolved reference: TestSignedInt
Error:(75, 64) Kotlin: Unresolved reference: TestSignedLong
Error:(84, 64) Kotlin: Unresolved reference: TestFixedInt
Error:(93, 64) Kotlin: Unresolved reference: TestDouble
Error:(102, 64) Kotlin: Unresolved reference: TestBoolean
Error:(122, 43) Kotlin: Unresolved reference: TestAllTypes
Error:(122, 58) Kotlin: Unresolved reference: TestAllTypes
Error:(158, 64) Kotlin: Unresolved reference: TestOuterMessage
Error:(180, 64) Kotlin: Unresolved reference: TestRepeatedIntMessage
Error:(191, 64) Kotlin: Unresolved reference: TestRepeatedObjectMessage
Error:(202, 44) Kotlin: Unresolved reference: TestEnum
Error:(202, 71) Kotlin: Unresolved reference: TestEnum
Error:(212, 44) Kotlin: Unresolved reference: TestMap
```

Same with JRE 11.

**Comment by reitzig** [comment]
The solution is to build through Gradle; IDEA doesn't build things correctly on its own, as it seems.

I was trying to run the test through the JUnit test runner; IDEA doesn't offer to create single-class/-method test configurations with the Gradle runner, by default. One has to change these settings:

![image](https://user-images.githubusercontent.com/4246780/49716058-cd7f7400-fc52-11e8-9287-c00f92d16bc1.png)

**Comment by sandwwraith** [comment]
For now, all projects with serialization require delegating to Gradle, as root readme says. This project also uses generated test data, which is automatically generated before running jvm tests via Gradle or by invoking Gradle task `generateTestProto`.

So it is usually sufficient to run just `gradlew build`: Gradle will automatically download 4.7 version, and generate everything

**Comment by reitzig** [comment]
It's not clear to me that the main README applies to developing this library itself; it seems to instruct the reader on how to _use_ it.

That said, I'm mainly leaving this hear to document the necessary steps for people Googling the same issues as I. Is there a better place for it?

(It _is_ mildly annoying that a "main" Kotlin project won't just build and run with the current IDEA. Even after taking above steps, the Gradle-run tests are not properly picked up by the test running toolbox in IDEA; this works in other Gradle projects for me. But all of that is, I guess more an issue for the IDEA tracker than this one?)

**Comment by sandwwraith** [comment]
Yes, it is mainly for library users, but the library itself is a serialization user (in tests) :) 

I think I'll move some points from our conversation to `docs/building.md` since there is obviously a lack of instructions.

Problem with IDEA run should be resolved after #265. However, for this particular project, you still have to pre-generate test data with Gradle and then point IDEA to the folder with generated stuff – AFAIK, JPS build can't automatically recognize such Gradle tasks.
